lang/cs/Org.Apache.REEF.IMRU/OnREEF/Driver/StateMachine/SystemStateMachine.cs

```diff
@@ -0,0 +1,91 @@
+    /// </summary>
+    internal sealed class SystemStateMachine
+    {
+        internal readonly static IDictionary<StateTransition<SystemState, SystemStateEvent>, SystemState> Transitions = 
+            new ReadOnlyDictionary<StateTransition<SystemState, SystemStateEvent>, SystemState>(
+                new Dictionary<StateTransition<SystemState, SystemStateEvent>, SystemState>()
+        {
+            { new StateTransition<SystemState, SystemStateEvent>(SystemState.WaitingForEvaluator, SystemStateEvent.AllContextsAreReady), SystemState.SubmittingTasks },
+            { new StateTransition<SystemState, SystemStateEvent>(SystemState.WaitingForEvaluator, SystemStateEvent.NotRecoverable), SystemState.Fail },
+            { new StateTransition<SystemState, SystemStateEvent>(SystemState.SubmittingTasks, SystemStateEvent.AllTasksAreRunning), SystemState.TasksRunning },
+            { new StateTransition<SystemState, SystemStateEvent>(SystemState.SubmittingTasks, SystemStateEvent.FailedNode), SystemState.ShuttingDown },
+            { new StateTransition<SystemState, SystemStateEvent>(SystemState.SubmittingTasks, SystemStateEvent.FailedNode), SystemState.ShuttingDown },
+            { new StateTransition<SystemState, SystemStateEvent>(SystemState.TasksRunning, SystemStateEvent.FailedNode), SystemState.ShuttingDown },
+            { new StateTransition<SystemState, SystemStateEvent>(SystemState.TasksRunning, SystemStateEvent.FailedNode), SystemState.ShuttingDown },
+            { new StateTransition<SystemState, SystemStateEvent>(SystemState.TasksRunning, SystemStateEvent.AllTasksAreCompleted), SystemState.TasksCompleted },
+            { new StateTransition<SystemState, SystemStateEvent>(SystemState.ShuttingDown, SystemStateEvent.Recover), SystemState.WaitingForEvaluator },
+            { new StateTransition<SystemState, SystemStateEvent>(SystemState.ShuttingDown, SystemStateEvent.NotRecoverable), SystemState.Fail }
+        });
+
+        private SystemState _currentState;
+
+        /// <summary>
+        /// create a new SysteState starting from WaitingForEvaluator
+        /// </summary>
+        internal SystemStateMachine()
+        {
+            _currentState = SystemState.WaitingForEvaluator;
+        }
+
+        /// <summary>
+        /// Returns current state
+        /// </summary>
+        internal SystemState CurrentState
+        {
+            get { return _currentState; }
+        }
+
+        /// <summary>
+        /// Based on the event and current state to determine the next state
+        /// </summary>
+        /// <param name="systemEvent"></param>
+        /// <returns></returns>
+        internal SystemState GetNext(SystemStateEvent systemEvent)
+        {
+            var transition = new StateTransition<SystemState, SystemStateEvent>(_currentState, systemEvent);
+            SystemState nextState;
+            if (!Transitions.TryGetValue(transition, out nextState))
```

Comment:
There are three types of pairs "current state - incoming event": 
1. event is valid in current state and changes state
2. event is valid in current state but doesn't change state (for example, `FailedNode` in `ShuttingDown` state)
3. event is not valid in current state (for example, `ShuttingDown` in `WaitingForEvaluator` state - we shouldn't get this event, since we don't have anything to shut down and are going straight to fail)

This state machine only covers the first case (and this if will throw exception if we get `FailedNode` in `ShuttingDown`). We probably need to distinguish between cases 2 and 3 at this level of code, since it will be more reliable than keeping track of this at event handler code (separately from state machine).